nexus/blueprint-execution/src/resource_allocation.rs

```diff
@@ -0,0 +1,998 @@
+            );
+        }
+
+        info!(
+            self.log, "external IP allocation required for zone";
+            "zone_type" => zone_type,
+            "zone_id" => %zone_id,
+            "ip" => %external_ip,
+        );
+
+        Ok(false)
+    }
+
+    // Helper function to determine whether a given NIC is already allocated to
+    // a specific service zone.
+    async fn is_nic_already_allocated(
+        &self,
+        zone_type: &'static str,
+        zone_id: Uuid,
+        nic: &NetworkInterface,
+    ) -> anyhow::Result<bool> {
+        let allocated_nics = self
+            .datastore
+            .service_list_network_interfaces(self.opctx, zone_id)
+            .await
+            .with_context(|| {
+                format!("failed to look up NICs for {zone_type} {zone_id}")
+            })?;
+
+        if !allocated_nics.is_empty() {
+            // All the service zones that want NICs only expect to have a single
+            // one. Bail out here if this zone already has one or more allocated
+            // NICs but not the one we think it needs.
+            //
+            // This doesn't check the allocated NIC's subnet against our NICs,
+            // because that would require an extra DB lookup. We'll assume if
+            // these main properties are correct, the subnet is too.
+            for allocated_nic in &allocated_nics {
+                if allocated_nic.ip.ip() == nic.ip
+                    && *allocated_nic.mac == nic.mac
+                    && allocated_nic.slot == i16::from(nic.slot)
+                    && allocated_nic.primary == nic.primary
+                {
+                    info!(
+                        self.log, "found already-allocated NIC";
+                        "zone_type" => zone_type,
+                        "zone_id" => %zone_id,
+                        "nic" => ?allocated_nic,
+                    );
+                    return Ok(true);
+                }
+            }
+
+            warn!(
+                self.log, "zone has unexpected NICs allocated";
+                "zone_type" => zone_type,
+                "zone_id" => %zone_id,
+                "want_nic" => ?nic,
+                "allocated_nics" => ?allocated_nics,
+            );
+
+            bail!(
+                "zone {zone_id} already has {} non-matching NIC(s) allocated",
+                allocated_nics.len()
+            );
+        }
+
+        info!(
+            self.log, "NIC allocation required for zone";
+            "zone_type" => zone_type,
+            "zone_id" => %zone_id,
+            "nid" => ?nic,
+        );
+
+        Ok(false)
+    }
+
+    // Nexus and ExternalDns both use non-SNAT service IPs; this method is used
+    // to allocate external networking for both of them.
+    async fn ensure_external_service_ip(
+        &self,
+        zone_type: &'static str,
+        service_id: Uuid,
+        external_ip: IpAddr,
+        ip_name: &Name,
+    ) -> anyhow::Result<()> {
```

Comment:
Thoughts on `use anyhow::Result`?

Reply:
No objection, I just have a light preference for this because when I see `Result<()>` I have to either remember or look up which typealias it is.